src/formula/variant.cpp

```diff
@@ -522,119 +412,99 @@ int variant::as_decimal() const
 }
 
-const std::map<variant,variant>& variant::as_map() const
+const std::map<variant, variant>& variant::as_map() const
 {
-	must_be(TYPE_MAP);
-	assert(map_);
-	return *map_;
+	must_be(VARIANT_TYPE::TYPE_MAP);
+	return value_cast<game_logic::variant_map>()->get_container();
 }
 
 variant variant::operator+(const variant& v) const
 {
-	if(type_ == TYPE_LIST) {
-		if(v.type_ == TYPE_LIST) {
-			std::vector<variant> res;
-			res.reserve(list_->size() + v.list_->size());
-			for(size_t i = 0; i<list_->size(); ++i) {
-				const variant& var = (*list_)[i];
-				res.push_back(var);
-			}
+	if(is_list() && v.is_list()) {
+		auto& list = value_cast<game_logic::variant_list>()->get_container();
+		auto& other_list = v.value_cast<game_logic::variant_list>()->get_container();
 
-			for(size_t j = 0; j<v.list_->size(); ++j) {
-				const variant& var = (*v.list_)[j];
-				res.push_back(var);
-			}
+		std::vector<variant> res;
+		res.reserve(list.size() + other_list.size());
+
+		for(const auto& member : list) {
+			res.push_back(member);
+		}
 
-			return variant(&res);
+		for(const auto& member : other_list) {
+			res.push_back(member);
 		}
+
+		return variant(&res);
 	}
-	if(type_ == TYPE_MAP) {
-		if(v.type_ == TYPE_MAP) {
-			std::map<variant,variant> res(*map_);
 
-			for(std::map<variant,variant>::const_iterator i = v.map_->begin(); i != v.map_->end(); ++i) {
-				res[i->first] = i->second;
-			}
+	if(is_map() && v.is_map()) {
+		std::map<variant, variant> res = value_cast<game_logic::variant_map>()->get_container();
 
-			return variant(&res);
+		for(const auto& member : v.value_cast<game_logic::variant_map>()->get_container()) {
+			res[member.first] = member.second;
 		}
+
+		return variant(&res);
 	}
-	if(type_ == TYPE_DECIMAL || v.type_ == TYPE_DECIMAL) {
-		return variant( as_decimal() + v.as_decimal() , DECIMAL_VARIANT);
+
+	if(is_decimal() || v.is_decimal()) {
+		return variant(as_decimal() + v.as_decimal() , DECIMAL_VARIANT);
 	}
 
 	return variant(as_int() + v.as_int());
 }
 
 variant variant::operator-(const variant& v) const
 {
-	if(type_ == TYPE_DECIMAL || v.type_ == TYPE_DECIMAL) {
-		return variant( as_decimal() - v.as_decimal() , DECIMAL_VARIANT);
+	if(is_decimal() || v.is_decimal()) {
+		return variant(as_decimal() - v.as_decimal() , DECIMAL_VARIANT);
 	}
 
 	return variant(as_int() - v.as_int());
 }
 
 variant variant::operator*(const variant& v) const
 {
-	if(type_ == TYPE_DECIMAL || v.type_ == TYPE_DECIMAL) {
+	if(is_decimal() || v.is_decimal()) {
 
 		long long long_int = as_decimal();
 
 		long_int *= v.as_decimal();
 
 		long_int /= 100;
 
-		if( long_int%10 >= 5) {
+		if(long_int%10 >= 5) {
 			long_int /= 10;
 			++long_int;
-		} else
+		} else {
 			long_int/=10;
+		}
 
-		return variant( static_cast<int>(long_int) , variant::DECIMAL_VARIANT );
+		return variant(static_cast<int>(long_int) , variant::DECIMAL_VARIANT );
 	}
 
 	return variant(as_int() * v.as_int());
 }
 
 variant variant::operator/(const variant& v) const
```

Comment:
Seems like the multiplication and division implementations for decimals could use some work...

(Nothing to do with the variant refactor, I just happened to notice it.)